File: EpipolarGeometryAndStereoVision/stereo.cpp
#include <opencv2/opencv.hpp>
#include "opencv2/core/core.hpp"
#include "opencv2/calib3d/calib3d.hpp"
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc/imgproc.hpp>
#include <stdio.h>
#include <string.h>

int main(int argc, char** argv)
{

  cv::Mat imgL,imgR;
  
  imgL = cv::imread("images/im0.png",0);
  cv::resize(imgL,imgL,cv::Size(600,600));
  imgR = cv::imread("images/im1.png",0);
  cv::resize(imgR,imgR,cv::Size(600,600));
  
  int minDisparity = 0;
  int numDisparities = 64;
  int blockSize = 8;
  int disp12MaxDiff = 1;
  int uniquenessRatio = 10;
  int speckleWindowSize = 10;
  int speckleRange = 8;

  cv::Ptr<cv::StereoSGBM> stereo = cv::StereoSGBM::create(minDisparity,numDisparities,blockSize,disp12MaxDiff,uniquenessRatio,speckleWindowSize,speckleRange);

  cv::Mat disp;
  stereo->compute(imgL,imgR,disp);
  
  cv::normalize(disp, disp, 0, 255, cv::NORM_MINMAX, CV_8UC1);  
  
  cv::imshow("Left image",imgL);
  cv::imshow("Right image",imgR);
  cv::imshow("disparity",disp);
  cv::waitKey(0);

  return 0;
}

File: EpipolarGeometryAndStereoVision/stereo.py
from __future__ import print_function

import numpy as np
import cv2

def main():
    imgL = cv2.imread("images/im0.png",0)
    imgL = cv2.resize(imgL,(600,600))

    imgR = cv2.imread("images/im1.png",0)
    imgR = cv2.resize(imgR,(600,600))

    # Setting parameters for StereoSGBM algorithm
    minDisparity = 0
    numDisparities = 64
    blockSize = 8
    disp12MaxDiff = 1
    uniquenessRatio = 10
    speckleWindowSize = 10
    speckleRange = 8

    # Creating an object of StereoSGBM algorithm
    stereo = cv2.StereoSGBM_create(minDisparity = minDisparity,
        numDisparities = numDisparities,
        blockSize = blockSize,
        disp12MaxDiff = disp12MaxDiff,
        uniquenessRatio = uniquenessRatio,
        speckleWindowSize = speckleWindowSize,
        speckleRange = speckleRange
    )

    # Calculating disparith using the StereoSGBM algorithm
    disp = stereo.compute(imgL, imgR).astype(np.float32)

    # Calculating disparith using the StereoSGBM algorithm
    disp = cv2.normalize(disp,0,255,cv2.NORM_MINMAX)
    
    # Displaying the disparity map
    cv2.imshow("disparity",disp)
    cv2.imshow("left image",imgL)
    cv2.imshow("right image",imgR)
    cv2.waitKey(0)
    

if __name__ == '__main__':
    main()
    cv2.destroyAllWindows()
